File: leetcode/20-valid-parentheses/main.go
package main

import (
	"fmt"
)

func isValid(s string) bool {
	parentheses := map[string]string{
		"}": "{",
		"]": "[",
		")": "(",
	}
	var stack []string
	for i := 0; i < len(s); i++ {
		curr := string(s[i])
		if curr == "{" || curr == "[" || curr == "(" {
			stack = append(stack, curr)
		} else if len(stack) == 0 {
			return false
		} else if parentheses[curr] != stack[len(stack)-1] {
			return false
		} else {
			stack = stack[:len(stack)-1]
		}
	}
	return len(stack) == 0
}

func main() {
	fmt.Println(isValid(""))
	fmt.Println(isValid("()"))
	fmt.Println(isValid("("))
	fmt.Println(isValid(")"))
	fmt.Println(isValid("(}"))
	fmt.Println(isValid("([)]"))
	fmt.Println(isValid("{[]}"))
	fmt.Println(isValid("{[[]}"))
}

File: leetcode/20-valid-parentheses/main.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s) == 0:
            return True
        stack = []
        m = {
            "}": "{",
            "]": "[",
            ")": "(",
        }
        for c in s:
            if c == "{" or c == "[" or c == "(":
                stack.append(c)
            elif c in m:
                if len(stack) > 0 and stack[-1] == m[c]:
                    stack.pop()
                else:
                    return False
        return len(stack) == 0


"""
    simplified version: only '('
"""


class Solution(object):
    def isValid(self, s):
        if len(s) == 0:
            return True
        openCount = 0
        for c in s:
            if c == '(':
                openCount += 1
            else:
                if openCount == 0:
                    return False
                else:
                    openCount -= 1
        return openCount == 0


print(Solution().isValid("(())"))
print(Solution().isValid("(()"))
print(Solution().isValid("(())()"))
